test/Groups/forms.jl

```diff
@@ -323,3 +323,280 @@ end
    is_true,z = iscongruent(Q1,Q2)
    @test !is_true
 end
+
+# TODO
+@testset "Relationship group - forms" begin
+   G = GL(6,3)
+   Op = GO(1,6,3)
+   Om = GO(-1,6,3)
+   F = base_ring(G)
+   B = zero_matrix(F,6,6)
+   for i in 1:6 B[i,7-i]=1 end
+   f = symmetric_form(B)
+   H = isometry_group(f)
+   @testset for x in gens(H)
+      @test f^x==f
+   end
+   @test order(H)==order(Op)
+   insert_block!(B,identity_matrix(F,2),3,3)
+   f = symmetric_form(B)
+   H = isometry_group(f)
+   @testset for x in gens(H)
+      @test f^x==f
+   end
+   @test order(H)==order(Om)
+
+   G = GL(4,2)
+   Op = GO(1,4,2)
+   Om = GO(-1,4,2)
+   F = base_ring(G)
+   B = zero_matrix(F,4,4)
+   B[1,4]=1; B[2,3]=1
+   f = quadratic_form(B)
+   H = isometry_group(f)
+   @testset for x in gens(H)
+      @test f^x==f
+   end
+   @test isconjugate(G,H,Op)[1]
+   B[2,2]=1; B[3,3]=1;
+   f = quadratic_form(B)
+   H = isometry_group(f)
+   @testset for x in gens(H)
+      @test f^x==f
+   end
+   @test isconjugate(G,H,Om)[1]
+   Q = f
+   f = corresponding_bilinear_form(Q)
+   H = isometry_group(f)
+   @testset for x in gens(H)
+      @test f^x==f
+   end
+   @test isconjugate(G,H,Sp(4,2))[1]
+
+   G = GL(5,9)
+   F = base_ring(G)
+   B = diagonal_matrix(F.([1,1,1,2,2]))
+   B[4,5] = gen(F)
+   B[5,4] = gen(F)^3
+   f = hermitian_form(B)
+   H = isometry_group(f)
+   @testset for x in gens(H)
+      @test f^x==f
+   end
+   @test order(H)==order(GU(5,3))
+
+   G = GU(2,3)
+   L = invariant_sesquilinear_forms(G)
+   @testset for f in L
+       for g in gens(G)
+          @test g.elm*f*conjugate_transpose(g.elm)==f
+       end
+   end
+   G = GO(-1,4,3)
+   L = invariant_bilinear_forms(G)
+   @testset for f in L
+       for g in gens(G)
+          @test g.elm*f*transpose(g.elm)==f
+       end
+   end
+   L = invariant_quadratic_forms(G)
+   @testset for m in L
+       Q = quadratic_form(m)
+       for g in gens(G)
+          @test Q^g==Q
```

Comment:
This check confuses me. The condition on `g` preserving the quadratic form given by `Q` should be "for all row vectors `v`, we have `v.Q.transpose(v) = (v.g).Q.transpose(v.g)` which in characteristic different from 2 is equivalent to `Q = g.Q.transpose(g)`. For a check that also works in char 2, see https://github.com/gap-system/gap/pull/4361/files

Now, if g is orthogonal with respect to the "standard" form, then `transpose(g) = inv(g)` and then the above check kinda makes sense; but not in general? Or am I missing something?

Anyway, I suggest we add (in a future PR!) a function like `preserves_form(form, matrix)` or `respects_form` or so, with methods for the various form types. Then the above line can become the following (and similar for the other tests)
```suggestion
          @test preserves_form(Q, g)
```

Reply:
to clarify the last part of the above comment: I am imaging here that we have types `QuadracticForm`, `BilinearForm`, `SesquilinearForm`, which store a matrix, and possibly a ring involution; then one can have methods like
```julia
preserves_form(f::BilinearForm, g) = g.elm*f*transpose(g.elm)==f
```
I think this would be useful to have anyway; one could then also apply the forms quite nicely:
```julia
Q = QuadraticForm(...)
v = rand(vectorspace)
z = Q(v) # evaluate the form
f = SesquilinearForm(...)
w = rand(vectorspace)
z2 = f(v, w)
```

Reply:
> This check confuses me. The condition on `g` preserving the quadratic form given by `Q` should be "for all row vectors `v`, we have `v.Q.transpose(v) = (v.g).Q.transpose(v.g)` which in characteristic different from 2 is equivalent to `Q = g.Q.transpose(g)`. For a check that also works in char 2, see https://github.com/gap-system/gap/pull/4361/files
> 
> Now, if g is orthogonal with respect to the "standard" form, then `transpose(g) = inv(g)` and then the above check kinda makes sense; but not in general? Or am I missing something?

I don't get exactly what you mean. Anyway, for quadratic forms, the condition  `v.Q.transpose(v) = (v.g).Q.transpose(v.g)` for every vector `v` does not imply the identity between matrices `Q = g.Q.transpose(g)`; rather, it implies that `Q - g.Q.transpose(g)` is a skew-symmetric matrix, but not necessarily a zero matrix (`X` is skew-symmetric iff `X = -transpose(X)` and `X` has zeros on the diagonal). This is all implemented in the definition of the method `==` for forms, and it holds in either odd and even characteristic.